Add some (matrix) functions for UniformScaling

ZeppLu opened the pull request:
Notations used here:
- I: identity matrix, known as UniformScaling in julia
- x: scalar
- v: column vector
- M: matrix

Methods list:
- [x] `x .\ I`
- [x] `I ^ x, I .^ x`
- [x] `exp, log, (a)sin/cos/tan/csc/sec/cot(h)`
- [x] `sqrt`
- [x] `v * I, v / I`
- [x] `real, imag`
- [x] `isposdef`
- [x] `tr` when `iszero(J.λ)`
- [ ] matrix decompositions
- [ ] broadcast functions of which 0 is a fixed point (so that matrix structure is preserved) (through `BroadcastStyle`?):
  - [ ] `real, imag, abs, abs2, angle, conj, sqrt, cbrt, ...`
  - [ ] `float, complex, big, rationalize, ...` and other casting functions
- [ ] or more?

ZeppLu committed:
α .\ I, and test

ZeppLu committed:
I ^ α, I .^ α, and tests

ZeppLu committed:
exp/log, sqrt/cbrt, and trigonometric functions

tests not yet written

ZeppLu committed:
cbrt(I) removed, since cube root of matrix is rarely needed

In fact none of other matrices in LinearAlgebra
specializes cbrt(), I saw sqrt() and then added
cbrt() as well without ever thinking.

ZeppLu committed:
tests for exp, log, trigonometric functions on \bbC

ZeppLu committed:
minor update

changes: M    = ...
     to: M(x) = ...
in order to make such conversion reusable in
following tests

ZeppLu committed:
tests for exp, log, trigonometric functions on \bbR

ZeppLu committed:
add tests for sqrt

I forgot it...

ZeppLu commented:
I'm having trouble understanding

https://github.com/JuliaLang/julia/blob/16dcabeaf58529f5bcf16c084e1dc4e8a54b1129/stdlib/LinearAlgebra/src/uniformscaling.jl#L175-L178

IIUC, line 175-176 is redundant, since `Union{Bidiagonal{T},AbstractTriangular{T}}` has already been covered by `AbstractMatrix`. Or is there any special cases that should be paid attention to?

ZeppLu committed:
v * I, and tests

ZeppLu committed:
fix tests

`J` is used by following tests, but I mistakenly changed its value

ZeppLu committed:
v / I, and tests

ZeppLu committed:
fix tests, seriously

I made v a scalar, forgive me

ZeppLu committed:
remove possibly redundant code to see if tests fail

`Union{Bidiagonal,AbstractTriangular} \ UniformScaling`
seems redundant because there's another method
`AbstractMatrix \ UniformScaling`

ZeppLu commented:
Just did a `git log -L` to find out [the first commit that adds this line](https://github.com/JuliaLang/julia/commit/e4c79b0bb2f28c7c0df7b01869404b7cab13e37b#diff-2053e27c93f0362bfb456f36c77bb232R70), which is done by @andreasnoack ! There surely are times a programmer forgets why he wrote a piece of code, and I come across the same feeling now and then, hahaha

ZeppLu committed:
real, imag, and tests

ZeppLu committed:
isposdef, and tests

also made `isapprox` a standalone testset,
because it's irrelevant to `istriu`, `issymmetric`, etc

ZeppLu committed:
tr, and tests

ZeppLu committed:
rearrange test code to make it more concise

ZeppLu committed:
reduce size of matrix used for testing to reduce test time

ZeppLu committed:
pinv, and tests

andreasnoack commented:
Hi ZeppLu. It's great that you are working on this. However, it would be great if you could push your changes a little less frequently since each time you push changes, it triggers new CI runs.

ZeppLu commented:
This is my first time working with a CI, apologize for any troubles I've made and thanks for your advice!

dalum commented:
Looks good!  Maybe you could also add `^(J::UniformScaling, M::Matrix) = exp!(log(J.λ)*M)`.  On that note, maybe `^(x::Number, M::Matrix) = exp!(log(x)*M)` should probably be added first?

ZeppLu commented:
Adding `^(x::Number, M::Matrix)` and `exp!` is beyond the purpose to this PR, which only focuses on UniformScaling. I'd suggest you open another PR.

BTW, I'm going to be quite busy this month (Sep), and will be back here as soon as I'm free again.

dkarrasch commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -141,23 +165,29 @@ function det(J::UniformScaling{T}) where T
 /(J1::UniformScaling, J2::UniformScaling) = J2.λ == 0 ? throw(SingularException(1)) : UniformScaling(J1.λ/J2.λ)
 /(J::UniformScaling, A::AbstractMatrix) = lmul!(J.λ, inv(A))
-/(A::AbstractMatrix, J::UniformScaling) = J.λ == 0 ? throw(SingularException(1)) : A/J.λ
+/(A::AbstractVecOrMat, J::UniformScaling) = J.λ == 0 ? throw(SingularException(1)) : A/J.λ
```
I think this should remain `AbstractMatrix` for the same reason as above.

dkarrasch commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -79,11 +81,15 @@ istril(::UniformScaling) = true
 issymmetric(::UniformScaling) = true
 ishermitian(J::UniformScaling) = isreal(J.λ)
 
+isposdef(J::UniformScaling) = isreal(J.λ) && J.λ > 0
```
How about
```
isposdef(J::UniformScaling) = isposdef(J.λ)
```
?

ZeppLu replied:
you're right!

ZeppLu commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -141,23 +165,29 @@ function det(J::UniformScaling{T}) where T
 *(B::BitArray{2}, J::UniformScaling) = *(Array(B), J::UniformScaling)
 *(J::UniformScaling, B::BitArray{2}) = *(J::UniformScaling, Array(B))
-*(A::AbstractMatrix, J::UniformScaling) = A*J.λ
+*(A::AbstractVecOrMat, J::UniformScaling) = A*J.λ
```
Sorry my linear algebra class was not taught in English, I can't get what "the λ-fold of the identity matrix" means, could you explain it in detail?

mbauman replied:
It's just saying that it's some identity matrix multiplied by λ.

This operation, multiplying a `vec*mat` is frequently an error.  I typically think of an identity matrix as being larger than 1x1, so I initially agreed.  But this is well-defined if J is considered to be 1x1... and we do say it automatically infers its size based upon the operation at hand.

I do think some extra consideration is still needed here, though, as multiplying `rand(5) * fill(2, 1, 1)` yields a 5×1 matrix and not a 5-element vector (as it would return as it's currently defined).

dkarrasch replied:
Yes, the issue is the distinction between a simple vector ``v`` and the linear map ``x \mapsto v*x``. In the first case, `v` should be a vector as we know it. In the second, `v` should be an nx1 matrix. Take your example `rand(5) * fill(2, 1, 1)`. As it stands, it is non-sensical. Checking with `@which` what is actually done leads to

https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/stdlib/LinearAlgebra/src/matmul.jl#L62

You see, to perform the product you first need to reshape the vector into a matrix, and only then multiplication works as expected. This and the lines above the quote are somewhat strange "convenience" functions, I would say. Anyway, if we wanted to be consistent with what we already have, we should have
```
*(A::AbstractMatrix, J::UniformScaling) = A*J.λ
*(a::AbstractVector, J::UniformScaling) = reshape(a*J.λ, length(a), 1)
```

mbauman replied:
You'll probably enjoy the discussion in #4774 if you've not discovered that one yet.  In short, we've decided to allow vectors to participate in the linear algebra of matrices — and in many cases that means they behave as though they have an implicit trailing `x1` dimension.  This particular case is covered in https://github.com/JuliaLang/julia/pull/20423.  Sure it's implemented as a reshape, but the important part is the behavior — and this is the behavior we've decided to have.

dkarrasch replied:
Thanks @mbauman, for the link. I remember reading it some long time ago. I agree it's a design decision, and I thought that behavior is desired. Would
```
*(a::AbstractVector, J::UniformScaling) = reshape(a, length(a), 1) * J
```
be then more appropriate? We would leave the multiplication handling to more generic parts of the code.

ZeppLu replied:
Thanks you two! I also consider `*(a::AbstractVector, J::UniformScaling) = reshape(a, length(a), 1) * J` more reasonable

ZeppLu committed:
update according to reviewers' suggestions

- *(v::AbstractVector, J::UniformScaling)
- /(v::AbstractVector, J::UniformScaling)
- isposdef(J::UniformScaling)

ZeppLu committed:
Merge branch 'master' into I-methods

ZeppLu committed:
fix test

ZeppLu committed:
fix test (element-wise euqality)

dkarrasch commented on stdlib/LinearAlgebra/test/uniformscaling.jl:
```diff
@@ -82,24 +139,48 @@ end
+        @test all(v' / J .== v' / λ)
+        @test all(J \ v  .== λ \ v )
+        @test all(J \ v' .== λ \ v')
+    end
```
What's the reason for not having simply `v * J == v * λ`? Equality of vectors is defined as equality of all elements, so I see no reason to write this all out.

dkarrasch commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -106,11 +108,15 @@ issymmetric(::UniformScaling) = true
 ishermitian(J::UniformScaling) = isreal(J.λ)
 isposdef(J::UniformScaling) = isposdef(J.λ)
 
+isposdef(J::UniformScaling) = isposdef(J.λ)
```
This is defined already two lines above.

dkarrasch commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -227,6 +259,10 @@ Broadcast.broadcasted(::typeof(*), J::UniformScaling,x::Number) = UniformScaling
 
+Broadcast.broadcasted(::typeof(\), x::Number,J::UniformScaling) = UniformScaling(x\J.λ)
+
+Broadcast.broadcasted(::typeof(^), J::UniformScaling,x::Number) = UniformScaling(J.λ^x)
```
In the three lines above, maybe add a space between `,` and the respective last argument?

dkarrasch commented on stdlib/LinearAlgebra/src/uniformscaling.jl:
```diff
@@ -227,6 +259,10 @@ Broadcast.broadcasted(::typeof(*), J::UniformScaling,x::Number) = UniformScaling
+
+Broadcast.broadcasted(::typeof(^), J::UniformScaling,x::Number) = UniformScaling(J.λ^x)
+
 ==(J1::UniformScaling,J2::UniformScaling) = (J1.λ == J2.λ)
```
Same here: space in front of the second argument.

chriscoey commented:
hey @ZeppLu, thanks for doing this, any chance you could finish it off?

chriscoey commented:
also, could this fix the broadcasting with `I` issue at https://github.com/JuliaLang/julia/issues/23197?

dkarrasch committed:
Merge branch 'master' into I-methods

dkarrasch committed:
relax some multiplication/division tests

dkarrasch commented:
Anyone care to take a quick look? @jagot @JeffreySarnoff ?

jagot commented:
I think it looks fine!

vtjnash commented:
This introduced some build issues:
```
WARNING: Method definition isposdef(LinearAlgebra.UniformScaling{T} where T<:Number) in module LinearAlgebra at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:109 overwritten at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:111.
WARNING: Method definition ^(LinearAlgebra.UniformScaling{T} where T<:Number, Number) in module LinearAlgebra at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:118 overwritten at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:269.
WARNING: Method definition broadcasted(typeof(Base.:(^)), LinearAlgebra.UniformScaling{T} where T<:Number, Number) in module LinearAlgebra at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:267 overwritten at /data/vtjnash/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/uniformscaling.jl:272.
```

dkarrasch commented:
Sorry! Fix in #34819.
